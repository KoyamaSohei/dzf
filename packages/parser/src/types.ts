export enum Rrtype {
  CNAME,
  HINFO,
  MX,
  NS,
  PTR,
  SOA,
  TXT,
  A
}

export enum Class {
  IN,
  CS,
  CH,
  HS,
  NONE,
  ANY
}

export interface TTLState {
  ttl: number;
  isByDirective: boolean;
}

export interface RrHeader<T = Rrtype> {
  name: string;
  rrtype: T;
  class: Class;
  ttl: number;
}

export type RR = CNAME | HINFO | MX | NS | PTR | SOA | TXT | A;

export interface CNAME {
  hdr: RrHeader<Rrtype.CNAME>;
  target: string;
}

export interface HINFO {
  hdr: RrHeader<Rrtype.HINFO>;
  cpu: string;
  os: string;
}

export interface MX {
  hdr: RrHeader<Rrtype.MX>;
  preference: number;
  mx: string;
}

export interface NS {
  hdr: RrHeader<Rrtype.NS>;
  ns: string;
}

export interface PTR {
  hdr: RrHeader<Rrtype.PTR>;
  ptr: string;
}

export interface SOA {
  hdr: RrHeader<Rrtype.SOA>;
  ns: string;
  mbox: string;
  serial: number;
  refresh: number;
  retry: number;
  expire: number;
  minttl: number;
}

export interface TXT {
  hdr: RrHeader<Rrtype.TXT>;
  txt: string;
}

export interface A {
  hdr: RrHeader<Rrtype.A>;
  a: string;
}

export interface Lex {
  token: string;
  value: Tokenize;
  torc: Rrtype | Class;
  line: number;
  column: number;
}

export enum Tokenize {
  zEOF,
  zString,
  zBlank,
  zQuote,
  zNewline,
  zRrtype,
  zOwner,
  zClass,
  zDirOrigin,
  zDirTTL,

  zValue,
  zKey,

  zExpectOwnerDir,
  zExpectOwnerBl,
  zExpectAny,
  zExpectAnyNoClass,
  zExpectAnyNoClassBl,
  zExpectAnyNoTTL,
  zExpectAnyNoTTLBl,
  zExpectRrtype,
  zExpectRrtypeBl,
  zExpectRdata,
  zExpectDirTTLBl,
  zExpectDirTTL,
  zExpectDirOriginBl,
  zExpectDirOrigin
}
