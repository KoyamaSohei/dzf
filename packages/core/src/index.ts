import { Rrtype, Class } from './types';
export { Rrtype, Class };

export function isRrtype(t: string) {
  switch (t) {
    case 'CNAME':
      return Rrtype.CNAME;
    case 'HINFO':
      return Rrtype.HINFO;
    case 'MX':
      return Rrtype.MX;
    case 'NS':
      return Rrtype.NS;
    case 'PTR':
      return Rrtype.PTR;
    case 'SOA':
      return Rrtype.SOA;
    case 'TXT':
      return Rrtype.TXT;
    case 'A':
      return Rrtype.A;
    default:
      return null;
  }
}

export function toRrtype(t: Rrtype): string {
  switch (t) {
    case Rrtype.CNAME:
      return 'CNAME';
    case Rrtype.HINFO:
      return 'HINFO';
    case Rrtype.MX:
      return 'MX';
    case Rrtype.NS:
      return 'NS';
    case Rrtype.PTR:
      return 'PTR';
    case Rrtype.SOA:
      return 'SOA';
    case Rrtype.TXT:
      return 'TXT';
    case Rrtype.A:
      return 'A';
  }
}

export function isClass(c: string) {
  switch (c) {
    case 'IN':
      return Class.IN;
    case 'CS':
      return Class.CS;
    case 'CH':
      return Class.CH;
    case 'HS':
      return Class.HS;
    case 'NONE':
      return Class.NONE;
    case 'ANY':
      return Class.ANY;
    default:
      return null;
  }
}

export function toClass(c: Class) {
  switch (c) {
    case Class.IN:
      return 'IN';
    case Class.CS:
      return 'CS';
    case Class.CH:
      return 'CH';
    case Class.HS:
      return 'HS';
    case Class.NONE:
      return 'NONE';
    case Class.ANY:
      return 'ANY';
  }
}
