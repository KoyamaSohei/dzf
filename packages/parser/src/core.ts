import { ZoneLexer, ByteReader } from './lexer';
import { RrHeader, TTLState, Class, Rrtype, CNAME, HINFO, MX, NS, PTR, SOA, TXT, A, Tokenize, Lex, RR } from './types';

export class ZoneParser {
  private c: ZoneLexer;
  private origin: string;
  public file: string;
  private h: RrHeader | null;
  private defttl: TTLState | null;
  public constructor(r: ByteReader, origin: string, file: string, defaultTtl: number) {
    this.c = new ZoneLexer(r);
    this.origin = origin;
    this.file = file;
    this.h = null;
    this.defttl = {
      ttl: defaultTtl,
      isByDirective: true,
    }
  }

  public toAbsoluteName(name: string, origin: string) {
    if (origin == '') {
      if (name == '' || name == '@') {
        throw new SyntaxError(`invalid domain, name:${name} ,origin:${origin}`);
      }
      return name;
    }
    if (name == '@') {
      return origin
    }
    if (origin == '.') {
      return name + origin;
    }
    if (name.length > 0 && name[name.length - 1] == '.') {
      return name;
    }
    return name + '.' + origin;
  }

  public next(): RR | null {
    const zp = this;
    let state = Tokenize.zExpectOwnerDir;
    let h = (zp.h || {}) as RrHeader;
    for (let l = zp.c.next(); l.value !== Tokenize.zEOF; l = zp.c.next()) {
      switch (state) {
        case Tokenize.zExpectOwnerDir:
          {
            if (zp.defttl !== null) {
              h.ttl = zp.defttl.ttl;
            }
            h.class = Class.IN;
            switch (l.value) {
              case Tokenize.zNewline:
                {
                  state = Tokenize.zExpectOwnerDir;
                  break;
                }
              case Tokenize.zOwner:
                {
                  const n = this.toAbsoluteName(l.token, zp.origin);
                  h.name = n;
                  state = Tokenize.zExpectOwnerBl;
                  break;
                }
              case Tokenize.zDirTTL:
                {
                  state = Tokenize.zExpectDirTTLBl;
                  break;
                }
              case Tokenize.zDirOrigin:
                {
                  state = Tokenize.zExpectDirOriginBl;
                  break;
                }
              case Tokenize.zRrtype:
                {
                  h.rrtype = l.torc as Rrtype;
                  state = Tokenize.zExpectRdata;
                  break;
                }
              case Tokenize.zClass:
                {
                  h.class = l.torc as Class;
                  state = Tokenize.zExpectAnyNoClassBl;
                  break;
                }
              case Tokenize.zBlank:
                break;
              case Tokenize.zString:
                {
                  const t = parseInt(l.token);
                  if (isNaN(t)) {
                    throw new TypeError(`${l.token} is not number at ${l.line}: ${l.column}`);
                  }
                  h.ttl = t;
                  if (zp.defttl == null || !zp.defttl.isByDirective) {
                    zp.defttl = {
                      ttl: t,
                      isByDirective: false,
                    }
                  }
                  state = Tokenize.zExpectAnyNoTTLBl;
                  break;
                }
              default:
                {
                  throw new SyntaxError(`error at ${l.line}: ${l.column}`);
                }
            }
            break;
          }
        case Tokenize.zExpectDirTTLBl:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`no blank after $TTL-directive at ${l.line}:${l.column}`);
            }
            state = Tokenize.zExpectDirTTL;
            break;
          }
        case Tokenize.zExpectDirTTL:
          {
            if (l.value !== Tokenize.zString) {
              throw new SyntaxError(`expecting $TTL value at ${l.line}:${l.column}`);
            }
            const t = parseInt(l.token);
            if (isNaN(t)) {
              throw new SyntaxError(`${l.token} is not number at ${l.line}:${l.column}`);
            }
            zp.defttl = {
              ttl: t,
              isByDirective: true,
            }

            state = Tokenize.zExpectOwnerDir;
            break;
          }
        case Tokenize.zExpectDirOriginBl:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`no blank after $ORIGIN at ${l.line}:${l.column}`);
            }
            state = Tokenize.zExpectDirOrigin;
            break;
          }
        case Tokenize.zExpectDirOrigin:
          {
            if (l.value !== Tokenize.zString) {
              throw new SyntaxError(`expecting $ORIGIN at ${l.line}:${l.column}`);
            }
            zp.origin = l.token;
            state = Tokenize.zExpectOwnerDir;
            break;

          }
        case Tokenize.zExpectOwnerBl:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`no blank after owner at ${l.line}:${l.column}`);
            }
            state = Tokenize.zExpectAny;
            break;
          }
        case Tokenize.zExpectAny:
          {
            switch (l.value) {
              case Tokenize.zRrtype:
                {
                  if (zp.defttl == null) {
                    throw new SyntaxError(`missing TTL with no previous value at ${l.line}:${l.column}`);
                  }
                  h.rrtype = l.torc as Rrtype;
                  state = Tokenize.zExpectRdata;
                  break;
                }
              case Tokenize.zClass:
                {
                  h.class = l.torc as Class;
                  state = Tokenize.zExpectAnyNoClassBl;
                  break;
                }
              case Tokenize.zString:
                {
                  const t = parseInt(l.token);
                  if (isNaN(t)) {
                    throw new SyntaxError(`${l.token} is not number at ${l.line}:${l.column}`);
                  }
                  h.ttl = t;
                  if (zp.defttl === null || !zp.defttl.isByDirective) {
                    zp.defttl = {
                      ttl: t,
                      isByDirective: false,
                    }
                  }
                  state = Tokenize.zExpectAnyNoTTLBl;
                  break;
                }
              default:
                throw new SyntaxError(`expecting RR type,TTL ot Class at ${l.line}:${l.column}`);
            }
            break;
          }
        case Tokenize.zExpectAnyNoClassBl:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`no blank before class at ${l.line}:${l.column}`);
            }
            state = Tokenize.zExpectAnyNoClass;
            break;
          }
        case Tokenize.zExpectAnyNoTTLBl:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`no blank before ttl at ${l.line}:${l.column}`);
            }
            state = Tokenize.zExpectAnyNoTTL;
            break;
          }
        case Tokenize.zExpectAnyNoClass:
          {
            switch (l.value) {
              case Tokenize.zString:
                {
                  const t = parseInt(l.token);
                  if (isNaN(t)) {
                    throw new SyntaxError(`${l.token} is not number at ${l.line}:${l.column}`);
                  }
                  h.ttl = t;
                  if (zp.defttl === null || !zp.defttl.isByDirective) {
                    zp.defttl = {
                      ttl: t,
                      isByDirective: false
                    }
                  }
                  state = Tokenize.zExpectRrtypeBl;
                  break;
                }
              case Tokenize.zRrtype:
                {
                  h.rrtype = l.torc as Rrtype;
                  state = Tokenize.zExpectRdata;
                  break;
                }
              default:
                throw new SyntaxError(`expecting Rr type or ttl at ${l.line}:${l.column}`);
            }
            break;
          }
        case Tokenize.zExpectAnyNoTTL:
          {
            switch (l.value) {
              case Tokenize.zClass:
                {
                  h.class = l.torc as Class;
                  state = Tokenize.zExpectRrtypeBl;
                  break;

                }

              case Tokenize.zRrtype:
                {
                  h.rrtype = l.torc as Rrtype;
                  state = Tokenize.zExpectRdata;
                  break;
                }
              default:
                throw new SyntaxError(`expecting RR type or class at ${l.line}:${l.column}`);
            }
            break;
          }
        case Tokenize.zExpectRrtypeBl:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`expecting RR type or TTL at ${l.line}:${l.column}`);
            }
            state = Tokenize.zExpectRrtype;
            break;
          }
        case Tokenize.zExpectRrtype:
          {
            if (l.value !== Tokenize.zRrtype) {
              throw new SyntaxError(`unknown RR type at ${l.line}:${l.column}`);
            }
            h.rrtype = l.torc as Rrtype;
            state = Tokenize.zExpectRdata;
            break;
          }
        case Tokenize.zExpectRdata:
          {
            if (l.value !== Tokenize.zBlank) {
              throw new SyntaxError(`expecting blank at ${l.line}:${l.column}`);
            }
            return zp.parseRdata(h);
          }
      }
    }

    return null;
  }
  private parseRdata(hdr: RrHeader): RR | null {
    const zp = this;
    switch (hdr.rrtype) {
      case Rrtype.CNAME: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const target = zp.toAbsoluteName(l.token, zp.origin);
        if (target === null) {
          throw new SyntaxError(
            `${l.token} is not domain at ${l.line}:${l.column}`
          );
        }
        const rr: CNAME = {
          hdr: hdr as RrHeader<Rrtype.CNAME>,
          target
        };
        return rr;
      }
      case Rrtype.HINFO: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const cpu = l.token;
        const b = zp.c.next();
        if (b.value !== Tokenize.zBlank) {
          throw new SyntaxError(`expecting blank at ${b.line}:${b.column}`);
        }
        const lx = zp.c.next();
        if (lx.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting os string but got ${lx.token} at ${lx.line}: ${lx.column}`
          );
        }
        const os = lx.token;
        const rr: HINFO = {
          hdr: hdr as RrHeader<Rrtype.HINFO>,
          cpu,
          os
        };
        return rr;
      }
      case Rrtype.MX: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const preference = parseInt(l.token);
        if (isNaN(preference)) {
          throw new SyntaxError(
            `expecting mx preference,but ${l.token} is not number at ${l.line}:${l.column}`
          );
        }
        const b = zp.c.next();
        if (b.value !== Tokenize.zBlank) {
          throw new SyntaxError(`expecting blank at ${b.line}:${b.column}`);
        }
        const m = zp.c.next();
        if (m.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${m.token} at ${m.line}:${m.column}`
          );
        }
        const mx = zp.toAbsoluteName(m.token, zp.origin);
        if (mx === null) {
          throw new SyntaxError(
            `${m.token} is not domain at ${m.line}:${m.column}`
          );
        }
        const rr: MX = {
          hdr: hdr as RrHeader<Rrtype.MX>,
          preference,
          mx
        };
        return rr;
      }
      case Rrtype.NS: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const ns = zp.toAbsoluteName(l.token, zp.origin);
        if (ns === null) {
          throw new SyntaxError(
            `${l.token} is not domain at ${l.line}:${l.column}`
          );
        }
        const rr: NS = {
          hdr: hdr as RrHeader<Rrtype.NS>,
          ns
        };
        return rr;
      }
      case Rrtype.PTR: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const ptr = zp.toAbsoluteName(l.token, this.origin);
        if (ptr === null) {
          throw new SyntaxError(
            `${l.token} is not domain at ${l.line}:${l.column}`
          );
        }
        const rr: PTR = {
          hdr: hdr as RrHeader<Rrtype.PTR>,
          ptr
        };
        return rr;
      }
      case Rrtype.SOA: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const ns = zp.toAbsoluteName(l.token, zp.origin);
        if (ns === null) {
          throw new SyntaxError(
            `${l.token} is not domain at ${l.line}:${l.column}`
          );
        }
        const b = zp.c.next();
        if (b.value !== Tokenize.zBlank) {
          throw new SyntaxError(`expecting blank at ${b.line}:${b.column}`);
        }
        const lx = zp.c.next();
        const mbox = zp.toAbsoluteName(lx.token, zp.origin);
        if (mbox === null) {
          throw new SyntaxError(
            `${l.token} is not domain at ${l.line}:${l.column}`
          );
        }

        const rr: SOA = {
          ...(hdr as RrHeader<Rrtype.SOA>),
          ns,
          mbox,
          serial: 0,
          refresh: 0,
          retry: 0,
          expire: 0,
          minttl: 0
        };
        const b2 = zp.c.next();
        if (b2.value !== Tokenize.zBlank) {
          throw new SyntaxError(`expecting blank at ${b2.line}:${b2.column}`);
        }
        for (let i = 0; i < 5; i++) {
          const lx = zp.c.next();
          const n = parseInt(lx.token);
          if (isNaN(n)) {
            throw new SyntaxError(
              `${lx.token} is not number at ${lx.line}:${lx.column}`
            );
          }
          switch (i) {
            case 0: {
              rr.serial = n;
              const lx = zp.c.next();
              if (lx.value !== Tokenize.zBlank) {
                throw new SyntaxError(
                  `${lx.token} is not blank at ${lx.line}:${lx.column}`
                );
              }
              break;
            }
            case 1: {
              rr.refresh = n;
              const lx = zp.c.next();
              if (lx.value !== Tokenize.zBlank) {
                throw new SyntaxError(
                  `${lx.token} is not blank at ${lx.line}:${lx.column}`
                );
              }
              break;
            }
            case 2: {
              rr.retry = n;
              const lx = zp.c.next();
              if (lx.value !== Tokenize.zBlank) {
                throw new SyntaxError(
                  `${lx.token} is not blank at ${lx.line}:${lx.column}`
                );
              }
              break;
            }
            case 3: {
              rr.expire = n;
              const lx = zp.c.next();
              if (lx.value !== Tokenize.zBlank) {
                throw new SyntaxError(
                  `${lx.token} is not blank at ${lx.line}:${lx.column}`
                );
              }
              break;
            }
            case 4: {
              rr.minttl = n;

              break;
            }
          }
        }
        return rr;
      }
      case Rrtype.TXT: {
        const zp = this;
        const txt = zp.sliceRdata(zp.c);
        if (txt.length > 1) {
          throw new SyntaxError('arguments is too many');
        }
        const rr: TXT = {
          hdr: hdr as RrHeader<Rrtype.TXT>,
          txt: txt[0]
        };
        return rr;
      }
      case Rrtype.A: {
        const l = zp.c.next();
        if (l.value !== Tokenize.zString) {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
        const a = l.token;
        const rr: A = {
          hdr: hdr as RrHeader<Rrtype.A>,
          a
        };
        return rr;
      }
    }
  }

  private sliceRdata(c: ZoneLexer): string[] {
    const s: string[] = [];
    let quote = false;
    let empty = false;
    let l: Lex;
    for (
      l = c.next();
      l.value !== Tokenize.zNewline && l.value !== Tokenize.zEOF;
      l = c.next()
    ) {
      switch (l.value) {
        case Tokenize.zString: {
          empty = false;
          s.push(l.token);
          break;
        }
        case Tokenize.zBlank: {
          if (quote) {
            throw new SyntaxError(
              `not expected blank at ${l.line}:${l.column}`
            );
          }
          break;
        }
        case Tokenize.zQuote: {
          if (empty && quote) {
            s.push('');
          }
          quote = !quote;
          empty = true;
          break;
        }
        default: {
          throw new SyntaxError(
            `expecting string but got ${l.token} at ${l.line}:${l.column}`
          );
        }
      }
    }
    if (quote) {
      throw new SyntaxError(`missing \'"\' at ${l.line}:${l.column}`);
    }
    return s;
  }
}
