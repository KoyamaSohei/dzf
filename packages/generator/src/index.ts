import { RR, Rrtype, toClass, toRrtype } from '@dzfu/core';

export function genZone(rrs: RR[]): string {
  return rrs.reduce((acc, rr) => {
    let str;
    switch (rr.rrtype) {
      case Rrtype.CNAME:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.target}\n`;
        break;
      case Rrtype.HINFO:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.cpu}\t${rr.os}\n`;
        break;
      case Rrtype.MX:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.preference}\t${rr.mx}\n`;
        break;
      case Rrtype.NS:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.ns}\n`;
        break;
      case Rrtype.PTR:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.ptr}\n`;
        break;
      case Rrtype.SOA:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.ns}\t${rr.mbox}\t(\n\t${rr.serial}\n\t${rr.refresh}\n\t${
          rr.retry
        }\n\t${rr.expire}\n\t${rr.minttl}\n)\n`;
        break;
      case Rrtype.TXT:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.txt}\n`;
        break;
      case Rrtype.A:
        str = `${rr.name}\t${rr.ttl}\t${toClass(rr.class)}\t${toRrtype(
          rr.rrtype
        )}\t${rr.a}\n`;
        break;
    }
    return acc + str;
  }, '');
}
