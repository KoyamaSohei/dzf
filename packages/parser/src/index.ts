import { ZoneParser } from './core';
import { ByteReader } from './lexer';
export { Rrtype, RR, CNAME, HINFO, MX, NS, PTR, SOA, TXT, A, Class, } from './types';

export function parseZone(file: string, origin: string, defaultTtl: number = 3600) {
  if (file === '') return [];
  if (file.length > 0 && file[file.length - 1] !== '\n') {
    file += '\n';
  }
  const br = new ByteReader(file);
  const zp = new ZoneParser(br, origin, file, defaultTtl);
  const rrs = [];
  for (let res = zp.next(); res; res = zp.next()) {
    rrs.push(res);
  }
  return rrs;
}
