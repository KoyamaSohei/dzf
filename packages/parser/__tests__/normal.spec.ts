import { parseZone } from '../src';

describe('normal scenario', () => {

  describe('SOA record', () => {
    it('with class', () => {
      const p = parseZone('example1.com.\tIN\tSOA\tns1.example1.com.\troot.example1.com.\t( 10000 10000 10000 10000 10000 )\n'
        , 'example1.com.');
      expect(p).toMatchSnapshot();
    })
    it('with TTL', () => {
      const p = parseZone('example1.com.\t3600\tIN\tSOA\tns1.example1.com.\troot.example1.com.\t( 10000 10000 10000 10000 10000 )\n'
        , 'example1.com.');
      expect(p).toMatchSnapshot();
    })
    it('with TTL 2', () => {
      const p = parseZone('example1.com.\tIN\t3600\tSOA\tns1.example1.com.\troot.example1.com.\t( 10000 10000 10000 10000 10000 )\n'
        , 'example1.com.');
      expect(p).toMatchSnapshot();
    })
    it('without class', () => {
      const p = parseZone('example1.com.\tSOA\tns1.example1.com.\troot.example1.com.\t( 10000 10000 10000 10000 10000 )\n'
        , 'example1.com.');
      expect(p).toMatchSnapshot();
    })
    it('without class,with TTL', () => {
      const p = parseZone('example1.com.\t3600\tSOA\tns1.example1.com.\troot.example1.com.\t( 10000 10000 10000 10000 10000 )\n'
        , 'example1.com.');
      expect(p).toMatchSnapshot();
    })
  })

  describe('other rrtypes records', () => {
    it('CNAME with class', () => {
      const p = parseZone('foo.example1.com.\tCS\tCNAME example2.com.', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('CNAME with TTL', () => {
      const p = parseZone('foo.example1.com.\t3600\tCH\tCNAME example2.com.', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('HINFO', () => {
      const p = parseZone('example1.com.\tIN\tHINFO\tfoopc\tUNIX', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('MX', () => {
      const p = parseZone('example1.com.\t1800\tMX\t10\tmail1.example.com.', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('NS', () => {
      const p = parseZone('example1.com.\tHS\tNS\tns1.example1.com.', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('TXT', () => {
      const p = parseZone('example1.com.\tCS\tTXT\t"foo! foo! foo!"', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('A', () => {
      const p = parseZone('example1.com.\tANY\tA\t123.45.67.89', 'example1.com.');
      expect(p).toMatchSnapshot();
    })
  })

  describe('$ORIGIN', () => {
    it('basic', () => {
      const p = parseZone('$ORIGIN\texample1.com.\nexample1.com.\tANY\tA\t123.45.67.89', '');
      expect(p).toMatchSnapshot();
    })

    it('not override', () => {
      const p = parseZone('$ORIGIN\texample1.com.\nexample1.com.\tANY\tA\t123.45.67.89', 'foo.com.');
      expect(p).toMatchSnapshot();
    });

    it('@', () => {
      const p = parseZone('$ORIGIN\texample1.com.\n@\tANY\tA\t123.45.67.89', '');
      expect(p).toMatchSnapshot();
    })
  })

  describe('$TTL', () => {
    it('basic', () => {
      const p = parseZone('$TTL\t99999\nexample1.com.\tANY\tA\t123.45.67.89', 'example1.com.');
      expect(p).toMatchSnapshot();
    })

    it('not override', () => {
      const p = parseZone('$ORIGIN\texample1.com.\nexample1.com.\t3600\tANY\tA\t123.45.67.89', 'foo.com.');
      expect(p).toMatchSnapshot();
    });

  })

  describe('absolute domain', () => {
    it('name + origin', () => {
      const p = parseZone('dog \t3600\tANY\tA\t123.45.67.89', 'example1.com.');
      expect(p).toMatchSnapshot();
    })
  })

  describe('misc', () => {
    it('ignore comments', () => {
      const p = parseZone('dog \t3600\tANY\tA\t123.45.67.89; foo! foo!', 'example1.com.');
      expect(p).toMatchSnapshot();
    })
  })


})
