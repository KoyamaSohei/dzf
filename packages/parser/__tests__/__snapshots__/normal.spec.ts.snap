// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normal scenario $ORIGIN @ 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario $ORIGIN basic 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario $ORIGIN not override 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario $TTL basic 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "example1.com.",
      "rrtype": 7,
      "ttl": 99999,
    },
  },
]
`;

exports[`normal scenario $TTL not override 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario SOA record with TTL 1`] = `
Array [
  Object {
    "expire": 10000,
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 5,
      "ttl": 3600,
    },
    "mbox": "root.example1.com.",
    "minttl": 10000,
    "ns": "ns1.example1.com.",
    "refresh": 10000,
    "retry": 10000,
    "serial": 10000,
  },
]
`;

exports[`normal scenario SOA record with TTL 2 1`] = `
Array [
  Object {
    "expire": 10000,
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 5,
      "ttl": 3600,
    },
    "mbox": "root.example1.com.",
    "minttl": 10000,
    "ns": "ns1.example1.com.",
    "refresh": 10000,
    "retry": 10000,
    "serial": 10000,
  },
]
`;

exports[`normal scenario SOA record with class 1`] = `
Array [
  Object {
    "expire": 10000,
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 5,
      "ttl": 3600,
    },
    "mbox": "root.example1.com.",
    "minttl": 10000,
    "ns": "ns1.example1.com.",
    "refresh": 10000,
    "retry": 10000,
    "serial": 10000,
  },
]
`;

exports[`normal scenario SOA record without class 1`] = `
Array [
  Object {
    "expire": 10000,
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 5,
      "ttl": 3600,
    },
    "mbox": "root.example1.com.",
    "minttl": 10000,
    "ns": "ns1.example1.com.",
    "refresh": 10000,
    "retry": 10000,
    "serial": 10000,
  },
]
`;

exports[`normal scenario SOA record without class,with TTL 1`] = `
Array [
  Object {
    "expire": 10000,
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 5,
      "ttl": 3600,
    },
    "mbox": "root.example1.com.",
    "minttl": 10000,
    "ns": "ns1.example1.com.",
    "refresh": 10000,
    "retry": 10000,
    "serial": 10000,
  },
]
`;

exports[`normal scenario absolute domain name + origin 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "dog.example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario misc ignore comments 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "dog.example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario other rrtypes records A 1`] = `
Array [
  Object {
    "a": "123.45.67.89",
    "hdr": Object {
      "class": 5,
      "name": "example1.com.",
      "rrtype": 7,
      "ttl": 3600,
    },
  },
]
`;

exports[`normal scenario other rrtypes records CNAME with TTL 1`] = `
Array [
  Object {
    "hdr": Object {
      "class": 2,
      "name": "foo.example1.com.",
      "rrtype": 0,
      "ttl": 3600,
    },
    "target": "example2.com.",
  },
]
`;

exports[`normal scenario other rrtypes records CNAME with class 1`] = `
Array [
  Object {
    "hdr": Object {
      "class": 1,
      "name": "foo.example1.com.",
      "rrtype": 0,
      "ttl": 3600,
    },
    "target": "example2.com.",
  },
]
`;

exports[`normal scenario other rrtypes records HINFO 1`] = `
Array [
  Object {
    "cpu": "foopc",
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 1,
      "ttl": 3600,
    },
    "os": "UNIX",
  },
]
`;

exports[`normal scenario other rrtypes records MX 1`] = `
Array [
  Object {
    "hdr": Object {
      "class": 0,
      "name": "example1.com.",
      "rrtype": 2,
      "ttl": 1800,
    },
    "mx": "mail1.example.com.",
    "preference": 10,
  },
]
`;

exports[`normal scenario other rrtypes records NS 1`] = `
Array [
  Object {
    "hdr": Object {
      "class": 3,
      "name": "example1.com.",
      "rrtype": 3,
      "ttl": 3600,
    },
    "ns": "ns1.example1.com.",
  },
]
`;

exports[`normal scenario other rrtypes records TXT 1`] = `
Array [
  Object {
    "hdr": Object {
      "class": 1,
      "name": "example1.com.",
      "rrtype": 6,
      "ttl": 3600,
    },
    "txt": "foo! foo! foo!",
  },
]
`;
